ui/src/components/Timeline2/util.ts

```diff
@@ -0,0 +1,257 @@
+ */
+export class ImpreciseDate {
+  readonly year?: number;
+  readonly month?: number;
+  readonly day?: number;
+
+  constructor(raw?: string) {
+    if (!raw) {
+      return;
+    }
+
+    const yearRegex = /(?<year>\d{4})/.source;
+    const monthRegex = /(?<month>0?[1-9]|1[0-2])/.source;
+    const dayRegex = /(?<day>0?[1-9]|[1-2][0-9]|3[0-1])/.source;
+
+    const regex = new RegExp(
+      `^${yearRegex}(?:-${monthRegex}(?:-${dayRegex})?)?$`
+    );
+
+    const result = regex.exec(raw);
+    const groups = result?.groups;
+
+    if (!groups) {
+      return;
+    }
+
+    this.year = groups.year ? parseInt(groups.year, 10) : undefined;
+    this.month = groups.month ? parseInt(groups.month, 10) : undefined;
+    this.day = groups.day ? parseInt(groups.day, 10) : undefined;
+  }
+
+  isValid() {
+    if (!this.year) {
+      return false;
+    }
+
+    if (
+      this.month &&
+      this.day &&
+      endOfMonth(this.year, this.month) < this.day
+    ) {
+      return false;
+    }
+
+    return true;
+  }
+
+  getEarliest() {
+    if (!this.isValid() || !this.year) {
+      return;
+    }
+
+    return new Date(this.year, this.month ? this.month - 1 : 0, this.day || 1);
+  }
+
+  getLatest() {
+    if (!this.isValid() || !this.year) {
+      return;
+    }
+
+    const month = this.month || 12;
+    const day = this.day || endOfMonth(this.year, month);
+
+    return new Date(this.year, month - 1, day);
+  }
+}
+
+export function useTimelineKeyboardNavigation<T extends HTMLElement>(
+  items: Array<TimelineItem>,
+  onUnselect: () => void
+) {
+  const itemRefs = useMemo(() => items.map(() => createRef<T>()), [items]);
+
+  const onKeyDown = (event: KeyboardEvent<HTMLElement>) => {
+    if (event.key === 'Escape') {
+      onUnselect();
+      return;
+    }
+
+    const activeIndex = itemRefs.findIndex(
+      (ref) =>
+        ref.current === document.activeElement ||
+        ref.current?.contains(document.activeElement)
+    );
+
+    if (activeIndex < 0) {
+      return;
+    }
+
+    if (event.key === 'ArrowDown') {
+      const newIndex = Math.min(items.length - 1, activeIndex + 1);
+      itemRefs[newIndex].current?.focus();
+    }
+
+    if (event.key === 'ArrowUp') {
+      const newIndex = Math.max(0, activeIndex - 1);
+      itemRefs[newIndex].current?.focus();
+    }
+  };
+
+  return [itemRefs, { onKeyDown }] as const;
+}
+
+export function useTimelineItemKeyboardNavigation(
+  entity: Entity,
+  onSelect: (entity: Entity) => void
+) {
+  const onKeyUp = (event: KeyboardEvent<HTMLElement>) => {
+    if (event.key === ' ' || event.key === 'Enter') {
+      onSelect(entity);
+    }
+  };
+
+  return { onKeyUp } as const;
+}
+
+export function useEntitySuggestions(
+  schema: Schema,
+  fetchSuggestions: FetchEntitySuggestions
+) {
+  const [suggestions, setSuggestions] = useState<Array<Entity>>([]);
+  const [isFetching, setIsFetching] = useState(false);
+
+  // eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
When using the `useCallback` hook in React, we need to specify the dependencies (i.e. any variables or functions used in this hook) as a second argument. In case of `useCallback`, React guarantees that it will reuse the same function on rerenders as long as the dependencies do not change.

Not specifying all dependencies of the `useCallback` hook can lead to bugs, because the callback function may not be updated (but we’d expect it to update). That’s why we have a linter rule that checks that we always specify all dependencies when using `useCallback` (and other similar hooks).

So far so good. Our usage is a little special here, because we also use `throttle` from lodash, a helper that ensures that we use to ensure that we do not fetch suggestions more frequently than once per 200ms. The linter doesn’t know how exactly throttle works and what it returns, so it can’t be sure that this doesn’t violate the rule to always specify all hook dependencies.

We do however know that `throttle` is deterministic and the returned function will always behave in the exact same way as long as we pass it the same arguments. So I’ve ignored this specific linter rule in this specific case.